### DISCUSSION TITLE
How to Share Media Across Multiple Models Without Duplicating Files in Spatie Media Library?

### DISCUSSION THREAD

**Discussion by jehanzeb007** [spatie/laravel-medialibrary#3748] [Q&A]
I need to implement a feature where the same media file can be shared across multiple models without creating duplicate files in storage.  For example I have the previous row id and I will assign to new one in db and it will be child entry but in the storage I have only one file.
Here's the functionality I am aiming for:

Assigning Media to Multiple Models:
I want to link an existing media file to different model_type and model_id entries in the database without copying the physical file. The goal is to optimize storage usage by sharing the same file across models.

Deletion Behavior:

If the media is linked to only one model, deleting the media should remove both the database record and the physical file from storage.
If the media is linked to multiple models, deleting the media should only remove the database record for that model without affecting the shared file in storage.
Storage Optimization:
I want to ensure that storage usage is minimal by keeping a single copy of the file for all linked models while maintaining proper relationships in the database.

**Comment by timirey** [comment]
Did you find an answer?

**Comment by tareq-halaby** [comment]
Spatie Media Library doesn't natively support "shared" media across multiple models (one physical file, many DB records pointing to it), but there are **two solid approaches** depending on your needs.

---

## Approach 1: Copy Media Record + Share the Same File Path (Custom)

You can duplicate the `media` database record while keeping it pointing to the **same physical file** by manually creating a new media record with the same `disk`, `file_name`, and `uuid`, then symlinking or just copying the row.

However, this gets messy with Spatie's cleanup logic. The cleaner approach is below.

---

## Approach 2: Use `copyTo()` + Override Deletion (Recommended)

Spatie provides a `copyTo()` method that copies the media file and creates a new DB record on a different model. The downside is it **duplicates the file** on disk.

To avoid duplication while still linking, the real solution is a **reference/pivot model**:

### Step 1 — Designate one "owner" model that holds the actual media

```php
// The primary model that owns the real file
$post->addMedia($uploadedFile)->toMediaCollection('images');
$media = $post->getFirstMedia('images');
```

### Step 2 — Create a reference table to track shared media

```php
// Migration
Schema::create('model_media_references', function (Blueprint $table) {
    $table->id();
    $table->unsignedBigInteger('media_id'); // FK to spatie media table
    $table->morphs('referenceable');        // the secondary model
    $table->string('collection_name')->default('default');
    $table->timestamps();
});
```

```php
// ModelMediaReference model
class ModelMediaReference extends Model
{
    public function media(): BelongsTo
    {
        return $this->belongsTo(\Spatie\MediaLibrary\MediaCollections\Models\Media::class);
    }

    public function referenceable(): MorphTo
    {
        return $this->morphTo();
    }
}
```

### Step 3 — Add a trait to secondary models to resolve shared media

```php
trait HasSharedMedia
{
    public function sharedMediaReferences(): MorphMany
    {
        return $this->morphMany(ModelMediaReference::class, 'referenceable');
    }

    public function attachSharedMedia(int $mediaId, string $collection = 'default'): void
    {
        $this->sharedMediaReferences()->create([
            'media_id'        => $mediaId,
            'collection_name' => $collection,
        ]);
    }

    public function getSharedMedia(string $collection = 'default')
    {
        return $this->sharedMediaReferences()
            ->where('collection_name', $collection)
            ->with('media')
            ->get()
            ->pluck('media');
    }
}
```

### Step 4 — Use it

```php
// Original owner stores the file
$post->addMedia($file)->toMediaCollection('images');
$mediaId = $post->getFirstMedia('images')->id;

// Secondary model just references it — no file duplication!
$product->attachSharedMedia($mediaId, 'images');

// Retrieve it on the product
$media = $product->getSharedMedia('images');
$url = $media->first()->getUrl();
```

### Step 5 — Safe deletion logic

```php
// Override delete on ModelMediaReference to only delete the physical file
// when NO other references or the owner itself still exists

public static function booted(): void
{
    static::deleting(function (ModelMediaReference $ref) {
        $otherReferences = static::where('media_id', $ref->media_id)
            ->where('id', '!=', $ref->id)
            ->exists();

        $ownerStillExists = Media::find($ref->media_id)?->model !== null;

        // Only physically delete if no other references exist AND owner is gone
        if (!$otherReferences && !$ownerStillExists) {
            Media::find($ref->media_id)?->delete();
        }
    });
}
```

---

## Approach 3: Single Owner with `getUrl()` passed around (simplest)

If you just need the **URL** across models and don't need full Spatie media features on secondary models, simply store the `media_id` as a foreign key on secondary models and resolve through the original owner:

```php
// On your secondary model
public function sharedMedia(): BelongsTo
{
    return $this->belongsTo(
        \Spatie\MediaLibrary\MediaCollections\Models\Media::class,
        'shared_media_id'
    );
}

// Usage
$url = $product->sharedMedia->getUrl();
$thumb = $product->sharedMedia->getUrl('thumb');
```

This is the **least code** and works perfectly if you don't need `HasMedia` features on the secondary model.

---

## Summary

| Approach | File duplication | Spatie features on secondary | Complexity |
|---|---|---|---|
| `copyTo()` (native) | Yes | Full | Low |
| Reference table + trait | No | Partial (via custom) | Medium |
| BelongsTo `media_id` FK | No | Read-only URL | Very Low |

For your use case (link to child model, no file duplication, smart deletion), **Approach 2** is the right one. Hope this helps!
